### PR TITLE
Feature/php7.3 and other improvements

### PR DESCRIPTION
**Pull request by mtompset**
### Change
- Update PHP requirements to 7.3 or higher

### Fix
- Update PHPUnit to 9.x
- Code cleanup with transition to PHP 7.3
- Refactor Psr3Yii2Logger::interpolate to only return string

### Add
- Psr3Yii2Logger::interpolateArray to only return array
- Psr3Yii2Logger tests

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,15 +1,18 @@
-FROM silintl/php7:7.2
-MAINTAINER Mark Tompsett <mark_tompsett@sil.org>
+FROM php:7.3-apache-buster
+LABEL maintainer="Mark Tompsett <mark_tompsett@sil.org>"
```
Comment (Schparky) [inline Dockerfile:2]: Nice, I wasn't aware that `MAINTAINER` is deprecated.

File: Dockerfile [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
-
-# Fix timezone stuff from hanging.
-RUN apt-get update -y && echo "America/New_York" > /etc/timezone; \
-    apt-get install -y tzdata
+ENV REFRESHED_AT 2021-03-24
 
 # Make sure apt has current list/updates
+RUN apt-get update -y \
+# Fix timezone stuff from hanging.
+    && echo "America/New_York" > /etc/timezone \
```
Comment (Schparky) [inline Dockerfile:9]: This surprises me. Don't we typically use UTC?

File: src/LoggerBase.php [hunk 3]
```diff
@@ -16,27 +18,36 @@ abstract class LoggerBase extends \Psr\Log\AbstractLogger
...
+
+    /**
+     * Interpolate context values into the message placeholders.
+     *
+     * @param array $message The data to be logged.
+     * @param array $context (Optional:) The array of values to insert into the
+     *     corresponding placeholders.
+     * @return array The resulting log message.
+     */
+    protected function interpolateArray($message, array $context = []): array
```
Comment (forevermatt) [inline src/LoggerBase.php:46]: When the message is an array, we aren't actually interpolating the given context values into a message string. The array merge is simply returning an array with the combined data, so we could probably just get rid of this `interpolateArray()` method and directly do the `\array_merge()` wherever we were using this method.

File: src/LoggerBase.php [hunk 4]
```diff
@@ -16,27 +18,36 @@ abstract class LoggerBase extends \Psr\Log\AbstractLogger
...
+     * @return string The resulting log message.
      */
-    protected function interpolate($message, array $context = [])
+    protected function interpolate($message, array $context = []): string
     {
         if (is_string($message)) {
             return $this->interpolateString($message, $context);
         }
         
-        if (is_array($message) && $this->isArraySupported) {
```
Comment (forevermatt) [inline src/LoggerBase.php:30]: Since we're removing the general support for array messages, we can remove this `isArraySupported` property from  this LoggerBase class.

**Review by Schparky — APPROVED** [review]

File: src/Psr3Yii2Logger.php [hunk 5]
```diff
@@ -21,41 +22,37 @@ public function __construct()
...
-            case PsrLogLevel::EMERGENCY:
-                Yii::error($messageToLog);
-                break;
             case PsrLogLevel::ALERT:
-                Yii::error($messageToLog);
-                break;
             case PsrLogLevel::CRITICAL:
-                Yii::error($messageToLog);
-                break;
+            case PsrLogLevel::EMERGENCY:
```
Comment (forevermatt) [inline src/Psr3Yii2Logger.php:40]: Would you mind moving "EMERGENCY" back up to the order it was in? They are sorted by severity.

File: src/Psr3Yii2Logger.php [hunk 6]
```diff
@@ -21,41 +22,37 @@ public function __construct()
...
                 break;
             case PsrLogLevel::NOTICE:
-                Yii::info($messageToLog);
-                break;
             case PsrLogLevel::INFO:
                 Yii::info($messageToLog);
                 break;
             case PsrLogLevel::DEBUG:
-                Yii::trace($messageToLog);
+                Yii::debug($messageToLog);
```
Comment (forevermatt) [inline src/Psr3Yii2Logger.php:52]: Why the change from `Yii::trace()` to `Yii::debug()`?

File: tests/Psr3EveryLoggerTest.php [hunk 7]
```diff
@@ -97,4 +110,12 @@ public function testLogUnknownLogLevelYii2()
         $this->expectException('\Psr\Log\InvalidArgumentException');
         $logger->log('unknown', 'A test message');
     }
-}
\ No newline at end of file
+
+    public function testArrayYii2()
+    {
+        $logger = new Psr3Yii2Logger();
+        $errorMessages = ['Error Message 1', 'Error Message 2'];
```
Comment (forevermatt) [inline tests/Psr3EveryLoggerTest.php:117]: This is a subtle thing, but the array syntax is (as I understand it) for logging structured error data, not multiple different errors. This test obviously works, but implies that the array is for logging multiple errors, which could be misleading.

```suggestion
        $errorMessages = ['message' => 'There was a problem', 'details' => 'More info'];
```

File: src/Psr3Yii2Logger.php (hunk 6 above)
Comment (mtompset) [reply src/Psr3Yii2Logger.php:52]: trace is deprecated.

File: src/LoggerBase.php (hunk 3 above)
Comment (mtompset) [reply src/LoggerBase.php:46]: Done, not pushed yet.

File: src/LoggerBase.php (hunk 4 above)
Comment (mtompset) [reply src/LoggerBase.php:30]: Done, not pushed yet.

File: Dockerfile (hunk 1 above)
Comment (mtompset) [reply Dockerfile:2]: Discovered it recently while reading the Dockerfile specification.

File: Dockerfile (hunk 2 above)
Comment (mtompset) [reply Dockerfile:9]: Okay, I can change that. Not pushed yet.

File: src/Psr3Yii2Logger.php (hunk 5 above)
Comment (mtompset) [reply src/Psr3Yii2Logger.php:40]: When I did a code inspection, PHPStorm told me some weird messages. I'll try again.

File: tests/Psr3EveryLoggerTest.php (hunk 7 above)
Comment (mtompset) [reply tests/Psr3EveryLoggerTest.php]: I didn't even see your suggestion. I went a little more complicated.

File: tests/Psr3EveryLoggerTest.php [hunk 8]
```diff
@@ -97,4 +110,17 @@ public function testLogUnknownLogLevelYii2()
...
\ No newline at end of file
+
+    public function testArrayYii2()
+    {
+        $logger = new Psr3Yii2Logger();
+        $message = [
+            'from' => ['log@example.com'],
+            'to' => ['developer1@example.com', 'developer2@example.com'],
+            'subject' => 'Log message',
+            'formattedMessage' => ['Do not attempt to follow this example.'],
```
Comment (forevermatt) [inline tests/Psr3EveryLoggerTest.php:121]: Nice 😁

File: Dockerfile [hunk 9]
```diff
@@ -1,15 +1,18 @@
...
-
-# Fix timezone stuff from hanging.
-RUN apt-get update -y && echo "America/New_York" > /etc/timezone; \
-    apt-get install -y tzdata
+ENV REFRESHED_AT 2021-03-24
 
 # Make sure apt has current list/updates
+RUN apt-get update -y \
+# Fix timezone stuff from hanging.
+    && echo "Etc/UTC" > /etc/timezone \
```
Comment (forevermatt) [inline Dockerfile:9]: Nice change to UTC here 👍

**Review by forevermatt — APPROVED** [review]
